File: js/sr.js

const tf = require('@tensorflow/tfjs-node')
const fs = require('fs')
const jpeg = require('jpeg-js')


const readImage = async path => {
    const buf = fs.readFileSync(path)
    const pixels = jpeg.decode(buf, true)
    return pixels
}

const imageByteArray = async (image, numChannels) => {
    const pixels = image.data
    const numPixels = image.width * image.height;
    const values = new Int32Array(numPixels * numChannels);

    for (let i = 0; i < numPixels; i++) {
        for (let channel = 0; channel < numChannels; ++channel) {
            values[i * numChannels + channel] = pixels[i * 4 + channel];
        }
    }

    return values
}

const imageToInput = async (image, numChannels) => {
    const values = await imageByteArray(image, numChannels)
    const outShape = [1, image.height, image.width, numChannels]
    const input = tf.tensor4d(values, outShape, 'float32')

    return input
}


(async () => {
    const imagePath = 'input.jpg'
    const modelPath = './outputs/edsr_baseline_x4'



    let pixel_max_one = false
    if (modelPath.match(/carn|dbpn|esrgan|frsr|natsr|rrdb/)) {
        pixel_max_one = true
    }

    let channel_first = true
    if (modelPath.match(/4pp_eusr|eusr|frsr|natsr/)) {
        channel_first = false
    }

    try {
        const image = await readImage(imagePath)
        let input = await imageToInput(image, 3)

        if (channel_first) {
            input = tf.transpose(input, [0, 3, 1, 2])
        }
        if (pixel_max_one) {
            input = input / 255.0
        }

        const model = await tf.node.loadSavedModel(modelPath)
        const out = model.predict(input)

        // console.log(out)

        // let p = await out.arraySync()
        let p = out

        if (channel_first) {
            p = tf.transpose(p, [0, 2, 3, 1])
        }
        if (pixel_max_one) {
            p = p * 255.0
        }

        // console.log(p)

        const height = p.shape[1]
        const width = p.shape[2]
        const bufData = await p.buffer()

        const frameData = new Buffer(width * height * 4)
        for (let i = 0; i < height * width; i++) {
            for (let j = 0; j < 3; j++) {
                let v = bufData.values[i * 3 + j]
                v = Math.round(v)
                v = Math.max(0, v)
                v = Math.min(v, 255)
                frameData[i * 4 + j] = v
            }
            frameData[i * 4 + 3] = 0xff
        }

        const outImage = jpeg.encode({
            data: frameData,
            height,
            width,
        }, 100)
        fs.writeFileSync('out.jpg', outImage.data)
    } catch (e) {
        console.error(e)
    }
})()

File: py/sr.py
#!/usr/bin/env python
# coding: utf-8

import sys
import tensorflow as tf
import tensorflow_hub as hub
from PIL import Image
import numpy as np


def read_img(path):
    img = Image.open(path)
    arr = np.array(img)
    tfarr = tf.constant([arr])
    tfarr = tf.cast(tfarr, tf.float32)
    return tfarr


def predict(m, img, channel_first=True, pixel_max_one=False):

    if channel_first:
        img = tf.transpose(img, [0, 3, 1, 2])
    if pixel_max_one:
        img = img / 255.0

    pred = m(img)
    out = pred['out']

    if channel_first:
        out = tf.transpose(out, [0, 2, 3, 1])
    if pixel_max_one:
        out = out * 255.0

    out = out[0].numpy()
    img = Image.fromarray(np.round(np.clip(out, 0, 255)).astype(np.uint8), "RGB")
    return img


model_path = sys.argv[1]
img_path = sys.argv[2]
output_path = sys.argv[3]

channel_first = True
pixel_max_one = False

if 'carn' in model_path:
    pixel_max_one = True
if 'dbpn' in model_path:
    pixel_max_one = True
if 'esrgan' in model_path:
    pixel_max_one = True
if 'frsr' in model_path:
    pixel_max_one = True
if 'natsr' in model_path:
    pixel_max_one = True
if 'rrdb' in model_path:
    pixel_max_one = True

if '4pp_eusr' in model_path:
    channel_first = False
if 'eusr' in model_path:
    channel_first = False
if 'frsr' in model_path:
    channel_first = False
if 'natsr' in model_path:
    channel_first = False

print(model_path, img_path, output_path)
m = hub.KerasLayer(model_path, signature='serving_default', signature_outputs_as_dict=True)

img = read_img(img_path)
uimg = predict(m, img, channel_first=channel_first, pixel_max_one=pixel_max_one)
uimg.save(output_path)

